commitment/taproot.go

```diff
@@ -243,28 +293,17 @@ func TapBranchHash(preimage []byte) (*chainhash.Hash, error) {
 // TapLeafHash computes the TapHash of a TapLeaf node from its preimage
 // if possible, otherwise an error is returned.
 func TapLeafHash(preimage []byte) (*chainhash.Hash, error) {
-	// Empty preimage.
-	if len(preimage) == 0 {
-		return nil, ErrInvalidEmptyTapscriptPreimage
+	// Decode the script version and the script itself.
+	version, script, err := ScriptFromLeafPreimage(preimage)
+	if err != nil {
+		return nil, err
 	}
 
-	// Enforce that it is not including another Taproot Asset commitment.
-	if bytes.Contains(preimage, TaprootAssetsMarker[:]) {
+	// Verify that the script is not including a Taproot Asset commitment.
+	if IsTaprootAssetCommitmentScript(script) {
```

Comment:
I think ok as is if we want to keep the exiting control flow, but in the future I'd like to re-work the control flow to assert all the preconditions up front, then have this just be a dumb method that computes the tap leaf hash.

Reply:
agree.. Dumb method "TapLeafHash" and smart method which checks whatever validity constraints on the leaf

Reply:
Went with always enforcing these checks before computing the TapHash, as sometimes we're initializing `TapscriptPreimage` from a decoded Proof vs. using a safe constructor.